src/lib/storage/fsst_segment/fsst_encoder.hpp

```diff
@@ -0,0 +1,192 @@
+      total_length += value.size();
+      row_lengths.push_back(value.size());
+      // TODO(anybody): refactor the cast
+      row_pointers.push_back(reinterpret_cast<unsigned char*>(const_cast<char*>(value.data())));
+    }
+
+    // 7 + 2 * total_length -> See fsst_compress interface in fsst.h description.
+    compressed_values.resize(7 + 2 * total_length);
+
+    fsst_encoder_t* encoder = fsst_create(values.size(), row_lengths.data(), row_pointers.data(), 0);
+    [[maybe_unused]] size_t number_compressed_strings =
+        fsst_compress(encoder, values.size(), row_lengths.data(), row_pointers.data(), compressed_values.size(),
+                      compressed_values.data(), compressed_value_lengths.data(), compressed_value_pointers.data());
+
+    DebugAssert(number_compressed_strings == values.size(), "Compressed values buffer size was not large enough");
+
+    fsst_decoder_t decoder = fsst_decoder(encoder);
+    fsst_destroy(encoder);
+
+    return decoder;
+  }
+
+  uint64_t _create_offsets(pmr_vector<uint64_t>& compressed_value_lengths, pmr_vector<uint32_t>& offsets) {
+    size_t compressed_values_size = compressed_value_lengths.size();
+
+    // Calculate global offset based on compressed value lengths.
+    offsets[0] = 0;
+    uint64_t aggregated_offset_sum = 0;
+    for (size_t index{1}; index <= compressed_values_size; ++index) {
+      aggregated_offset_sum += compressed_value_lengths[index - 1];
+      offsets[index] = aggregated_offset_sum;
+    }
+
+    return aggregated_offset_sum;
+  }
+
+  uint64_t _create_reference_offsets(pmr_vector<uint32_t>& offsets, pmr_vector<uint64_t>& reference_offsets) {
+    auto reference_offsets_size = reference_offsets.size();
+    auto offsets_size = offsets.size();
+    // Since the first reference offset (which is always 0) is not saved,
+    // there are reference_offsets_size + 1 reference buckets.
+    if (offsets_size < reference_offsets_size + 1) {
+      // Do not use reference offsets.
+      return 0;
+    }
+    uint64_t n_elements_in_reference_bucket = static_cast<uint64_t>(
+        std::ceil(static_cast<double>(offsets.size()) / static_cast<double>(reference_offsets_size + 1)));
+
+    // Calculate start offset of each reference bucket.
+    for (size_t index = 0; index < reference_offsets_size; ++index) {
+      reference_offsets[index] = offsets[(index + 1) * n_elements_in_reference_bucket];
+    }
+
+    // Substract the reference offset from the original offset (create "zig-zag" pattern).
+    for (size_t index{n_elements_in_reference_bucket}; index < offsets_size; ++index) {
+      auto reference_offset_index = (index / n_elements_in_reference_bucket) - 1;
+
+      auto reference_offset = reference_offsets[reference_offset_index];
+      offsets[index] = offsets[index] - reference_offset;
+    }
+
+    return n_elements_in_reference_bucket;
+  }
+
+  template <typename T>
+  std::shared_ptr<AbstractEncodedSegment> _on_encode(const AnySegmentIterable<T> segment_iterable,
```

Comment:
You are copying AnySegmentIterable<T>, consider using a reference instead.

Reply:
This is the supplied interface that we merely implement. 

Dictionary:
https://github.com/hyrise/hyrise/blob/d50c5db23f1f53dfa5745706ea95cbc3b4f7a0a3/src/lib/storage/dictionary_segment/dictionary_encoder.hpp#L31-L32

LZ4 Encoder:
https://github.com/hyrise/hyrise/blob/d50c5db23f1f53dfa5745706ea95cbc3b4f7a0a3/src/lib/storage/lz4_segment/lz4_encoder.hpp#L64-L65